File: src/main/java/com/github/sadikovi/netflowlib/statistics/StatisticsWriter.java
/*
 * Copyright 2016 sadikovi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.sadikovi.netflowlib.statistics;

import java.io.IOException;
import java.nio.ByteOrder;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;

import org.apache.hadoop.fs.FSDataOutputStream;

/**
 * Interface to write NetFlow statistics as metadata to output stream provided. Handles stream
 * inside, e.g. `flush()` or `close()`, to ensure that resources are properly released after
 * reading.
 */
public class StatisticsWriter extends StatisticsAction {

  // writer buffer initial capacity
  public static final int INIT_BUF_CAPACITY = 128;

  /** Initialize statistics writer with output stream from creating a file */
  public StatisticsWriter(FSDataOutputStream out, ByteOrder order) throws IOException {
    if (out == null) {
      throw new IllegalArgumentException("Output stream is null");
    }

    if (order == null) {
      throw new IllegalArgumentException("Byte order is null");
    }

    this.out = out;
    this.order = order;

    checkStreamState();
  }

  /** StatisticsWriter with default BIG_ENDIAN byte order */
  public StatisticsWriter(FSDataOutputStream out) throws IOException {
    this(out, ByteOrder.BIG_ENDIAN);
  }

  /** Check stream non-empty state */
  private void checkStreamState() throws IOException {
    if (out.getPos() != 0L) {
      throw new IOException("Output stream is not empty");
    }
  }

  /**
   * Write metadata using Statistics object (recommended), as it does perform additional validation.
   *
   * @param stats Statistics to write
   * @return number of bytes written
   * @throws IOException
   */
  public long write(Statistics stats) throws IOException {
    return write(stats.getVersion(), stats.getCount(), stats.getOptions());
  }

  /**
   * Write metadata into output stream. This includes meta information for correct parsing, such as
   * magic numbers and byte order, header, and content of statistics options [[StatisticsOption]].
   *
   * @param version NetFlow version
   * @param count number of records
   * @param options list statistics options for certain fields
   * @return number of bytes written
   * @throws IOException
   */
  public long write(short version, long count, StatisticsOption[] options) throws IOException {
    checkStreamState();
    ByteBuf metabuf = null;

    if (options == null) {
      throw new IllegalArgumentException("Options array is null");
    }

    try {
      metabuf = Unpooled.directBuffer(INIT_BUF_CAPACITY).order(order);
      // write metadata statistics: magic number 1, magic number 2, byte order, each element takes
      // one byte of space, order independent
      metabuf.writeByte(MAGIC_NUMBER_1);
      metabuf.writeByte(MAGIC_NUMBER_2);
      metabuf.writeByte((byte)fromByteOrder(order));

      // write header (fixed size): length of header, version, and count of all the records
      int len = 10; // version 2 bytes and count 8 bytes
      metabuf.writeInt(len);
      metabuf.writeShort(version);
      metabuf.writeLong(count);

      // store full length of the options in metadata
      long fullLength = 0;
      for (StatisticsOption option: options) {
        if (option != null) {
          fullLength += STATISTICS_RECORD_SIZE;
        }
      }
      metabuf.writeLong(fullLength);

      // write options as TLV chain, we do not store length of the content
      byte[] optionArr = new byte[STATISTICS_RECORD_SIZE];
      ByteBuf optionBuf = Unpooled.wrappedBuffer(optionArr).order(order);

      for (StatisticsOption option: options) {
        if (option != null) {
          optionBuf.clear();
          // write field name
          optionBuf.writeInt((int)option.getField());
          // write field size
          short fieldSize = option.getSize();
          optionBuf.writeShort(fieldSize);
          // write min value based on size
          if (fieldSize == 1) {
            optionBuf.writeByte((byte)option.getMin());
            optionBuf.writeByte((byte)option.getMax());
          } else if (fieldSize == 2) {
            optionBuf.writeShort((short)option.getMin());
            optionBuf.writeShort((short)option.getMax());
          } else if (fieldSize == 4) {
            optionBuf.writeInt((int)option.getMin());
            optionBuf.writeInt((int)option.getMax());
          } else if (fieldSize == 8) {
            optionBuf.writeLong(option.getMin());
            optionBuf.writeLong(option.getMax());
          } else {
            throw new UnsupportedOperationException("Unsupported field size " + fieldSize);
          }

          metabuf.writeBytes(optionArr);
        }
      }

      // clear byte array and release option buffer
      if (optionBuf != null && optionBuf.refCnt() > 0) {
        optionBuf.release(optionBuf.refCnt());
      }
      optionArr = null;
      optionBuf = null;

      int writtenBytes = metabuf.readableBytes();
      metabuf.readBytes(out, writtenBytes);
      out.hflush();

      return (long)writtenBytes;
    } finally {
      if (metabuf != null && metabuf.refCnt() > 0) {
        metabuf.release(metabuf.refCnt());
      }
      metabuf = null;
      out.close();
    }
  }

  private FSDataOutputStream out = null;
}

File: src/main/scala/com/github/sadikovi/spark/netflow/index/StatisticsWriter.scala
/*
 * Copyright 2016 sadikovi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.sadikovi.spark.netflow.index

import java.io.OutputStream
import java.nio.ByteOrder

import scala.collection.GenTraversableOnce
import scala.collection.JavaConverters._

import io.netty.buffer.{ByteBuf, Unpooled}

import org.apache.hadoop.conf.{Configuration => HadoopConf}
import org.apache.hadoop.fs.{Path => HadoopPath}

/**
 * [[StatisticsWriter]] writes sequence of attributes into a file provided. Note that only numeric
 * types of attributes are supported. `OutputStream` is handled and closed automatically, with
 * underlying byte buffer.
 */
private[spark] class StatisticsWriter(
    private val endianness: ByteOrder,
    private val attrs: Seq[Attribute[_]]) {
  var buffer: ByteBuf = Unpooled.buffer().order(endianness)
  init()

  /** Write initial state: magic numbers, endianness */
  private def init(): Unit = {
    StatisticsUtils.withBytes(buffer, StatisticsUtils.INITIAL_STATE_SIZE) { buf =>
      buf.writeByte(StatisticsUtils.MAGIC_1)
      buf.writeByte(StatisticsUtils.MAGIC_2)
      buf.writeByte(StatisticsUtils.getEndiannessIndex(endianness))
    }
  }

  /** Write attribute header (flags, name, etc). Exposed to a package for testing purposes */
  private[index] def writeAttributeHeader(
      klass: Class[_], flags: Byte, key: String, hasNull: Boolean): Unit = {
    buffer.writeByte(StatisticsUtils.MAGIC_1)
    buffer.writeByte(StatisticsUtils.MAGIC_2)
    buffer.writeByte(flags)
    buffer.writeByte(StatisticsUtils.getBytes(klass))
    buffer.writeBoolean(hasNull)
    // Bytes of the key based on charset, this will write length as integer and array of bytes
    StatisticsUtils.writeValue(buffer, key, classOf[String])
  }

  /** Generic write method for arbitrary sequence of elements */
  private def writeElements(
      tpe: Byte,
      klass: Class[_], size: Int, elems: GenTraversableOnce[_], hasNull: Boolean): Unit = {
    buffer.writeByte(tpe)
    buffer.writeByte(StatisticsUtils.getBytes(klass))
    buffer.writeInt(size)
    val iter = elems.toIterator
    while (iter.hasNext) {
      val elem = iter.next
      if (hasNull) {
        buffer.writeBoolean(elem == null)
      }

      if (!hasNull || elem != null) {
        StatisticsUtils.writeValue(buffer, elem, klass)
      }
    }
  }

  private def writeElements(
      tpe: Byte, klass: Class[_], elems: GenTraversableOnce[_], hasNull: Boolean): Unit = {
    writeElements(tpe, klass, elems.size, elems, hasNull)
  }

  /** Internal method to write single attribute. Exposed to a package for testing purposes */
  private[index] def writeAttribute(attr: Attribute[_]): Unit = {
    val hasNull = attr.containsNull()
    writeAttributeHeader(attr.getClassTag(), attr.flags, attr.name, hasNull)
    // Attribute "count" parameter is always of long type regardless of runtime class
    attr.getCount() match {
      case Some(count) =>
        writeElements(StatisticsUtils.TYPE_COUNT, classOf[Long], Seq[Long](count), hasNull)
      case None => // do nothing
    }

    attr.getMinMax() match {
      case Some((min, max)) =>
        writeElements(StatisticsUtils.TYPE_MINMAX, attr.getClassTag(), Seq(min, max), hasNull)
      case None => // do nothing
    }

    attr.getSet() match {
      case Some(set) =>
        writeElements(StatisticsUtils.TYPE_SET, attr.getClassTag(), set.size(),
          set.iterator.asScala, hasNull)
      case None => // do nothing
    }
  }

  /** Internal save method, write each attribute and transfer bytes into output stream */
  private def internalSave(out: OutputStream): Unit = {
    for (attr <- attrs) {
      writeAttribute(attr)
    }
    buffer.getBytes(0, out, buffer.writerIndex)
  }

  /** Close underlying buffer */
  private def close(): Unit = try {
    buffer.release(buffer.refCnt)
  } finally {
    buffer = null
  }

  /** Save provided attributes into a file */
  def save(path: String, conf: HadoopConf = new HadoopConf(true)): Unit = {
    val pt = new HadoopPath(path)
    val fs = pt.getFileSystem(conf)
    val out = fs.create(pt, false)
    internalSave(out)
    out.close()
    close()
  }
}
